forutil add vectorized and scalar code

tang-hi opened the pull request:
### Description
ISSUE #12396

tang-hi committed:
add vectorized and scalar code

uschindler commented on lucene/core/src/java/org/apache/lucene/internal/vectorization/DefaultForUtil90.java:
```diff
@@ -0,0 +1,135 @@
+  /** Decode 128 integers into {@code longs}. */
+  public void decode(int bitsPerValue, DataInput in, int[] output) throws IOException {
+    for (int i = 0; i < 4 * bitsPerValue; ++i) {
+      tmp[i] = in.readInt();
```
We can use readInts() here

uschindler commented on lucene/core/src/java/org/apache/lucene/internal/vectorization/DefaultForUtil90.java:
```diff
@@ -0,0 +1,135 @@
+    }
+  }
+
+  /** Decode 128 integers into {@code longs}. */
```
comment wrong

ChrisHegarty commented:
This is starting to look much more like what I was expecting (but still a long way to go). Nice!

It looks like you @tang-hi brought in some code from [bitpacking][1], which is cool. Let's use this PR to collaborate further. Do you @tang-hi want to open branch-push access to me @ChrisHegarty (and whoever else desires to write code here)?


[1]: https://github.com/ChrisHegarty/bitpacking

uschindler commented:
> This is starting to look much more like what I was expecting (but still a long way to go). Nice!
> 
> It looks like you @tang-hi brought in some code from [bitpacking](https://github.com/ChrisHegarty/bitpacking), which is cool. Let's use this PR to collaborate further. Do you @tang-hi want to open branch-push access to me @ChrisHegarty (and whoever else desires to write code here)?

You should have push access to all PRs as PMC member. The contributor must acknowledge that.

tang-hi commented:
> Do you @tang-hi want to open branch-push access to me @ChrisHegarty (and whoever else desires to write code here)?

Of course. what should I do to open branch-push access?

uschindler commented:
> > Do you @tang-hi want to open branch-push access to me @ChrisHegarty (and whoever else desires to write code here)?
> 
> Of course. what should I do to open branch-push access?

We can automatically push to all contributors branches when they have a PR. You have confirmed that when opening the PR.

tang-hi commented:
Please feel free to submit your commits. I am a bit exhausted now and don't have the energy to look deeper. As for the issue with the failed tests, I believe the decode and encode functions are correct, although the scalar version of the code implementation is not ideal and was rushed. I haven't thoroughly tested encodeto32, so there might be some issues with it as well. To run the tests, you can use the following commands:
```
export CI=true
export RUNTIME_JAVA_HOME=/path/to/jdk20or21
./gradlew :lucene:core:test
```

uschindler commented:
It looks like scalar version passes tests (as GitHub uses java 17).

tang-hi commented:
Vectorized code is automatically generated, but I think we can manually write code for special bitPerValue (1, 2, 4, 8, 16) in the future to reduce code size. Of course, we can also handle the vector versions of 256 and 512 bits specially.

uschindler commented:
> Vectorized code is automatically generated, but I think we can manually write code for special bitPerValue (1, 2, 4, 8, 16) in the future to reduce code size. Of course, we can also handle the vector versions of 256 and 512 bits specially.

Yes, theres still room to autogenerate more code.

I'd also like to reduce the blind with those ever a d ever repeated long variable names. We can chain all call, I think.

But that has time till cleanup phase. I am good at rewriting code to look more compact without even understanding it. 😜

uschindler commented:
The reason why the backwards compatibility test ate failing is easy. We modified the Lucene90 codec and not created a new one.

The new code fails to read an index created with Lucene 9.0 or later. As we changed format we need a new codec. 

In my previous brach on which this is based on I just changed Lucene 9.0 codec.

I would ignore this. If Lucene core tests pass were fine for now. Creating the new codec should be done when we decided to merge this.

tang-hi committed:
format code style & fix comment error

tang-hi committed:
fix unthread safe problem

uschindler commented on lucene/core/src/java/org/apache/lucene/internal/vectorization/DefaultForUtil90.java:
```diff
@@ -0,0 +1,118 @@
+import org.apache.lucene.store.DataOutput;
+
+final class DefaultForUtil90 implements ForUtil90 {
+  private static final int[] tmp = new int[BLOCK_SIZE];
```
You can't make those static as this would break thread safety. To have tohose temporary buffers isolated, each caller gets its own instance of the ForUtil90, see `Provider#newForUtil90()`.

Please check the original code.

uschindler replied:
The same applies for the field below and those in Panama.

tang-hi committed:
add vectorized code error

ChrisHegarty commented on lucene/core/src/java20/org/apache/lucene/internal/vectorization/PanamaForUtil90.java:
```diff
@@ -0,0 +1,8994 @@
+  /** Encode 128 integers from {@code input} into {@code out}. */
+  @Override
+  public void encode(int[] input, int bitsPerValue, DataOutput out) throws IOException {
+    if (bitsPerValue == 32) {
```
If I'm not mistaken, then I don't think that we can ever get here with bitsPerValue == 32. If so, then this can be removed.

tang-hi replied:
Yes, you are correct. I mistakenly thought that we needed to compress bitsPerValue from 1 to 32.

ChrisHegarty commented on lucene/core/src/java20/org/apache/lucene/internal/vectorization/PanamaForUtil90.java:
```diff
@@ -0,0 +1,8994 @@
+  @Override
+  public void decode(int bitsPerValue, DataInput in, int[] output) throws IOException {
+    in.readInts(tmp, 0, 4 * bitsPerValue);
+    if (bitsPerValue == 32) {
```
(Same as above) - If I'm not mistaken, then I don't think that we can ever get here with bitsPerValue == 32. If so, then this can be removed.

tang-hi committed:
remove bitPerValue=32, we won't support this situation

ChrisHegarty committed:
Add TestForUtil90

ChrisHegarty committed:
Add a test that compares the scalar and vectorization code

tang-hi commented:
After running "./gradlew :lucene:core:generatePanamaForUtilInternal" and generating the .java file, the corresponding checksum value was not updated. Did I miss something? What command should I execute to update the corresponding checksum file?

uschindler commented:
You ran wrong task. It is called internal. 😜

Easiest to just run `gradlew :lucene:core:regenerate`

tang-hi committed:
update checksum

tang-hi commented:
Thank you, this worked. I think we can try using lucene-util to test the performance of this draft PR, as @jpountz  suggested in  #12396 . If we want to compare the vector and scalar versions, which script should we run in lucene-util?

uschindler commented:
> Thank you, this worked. I think we can try using lucene-util to test the performance of this draft PR, as @jpountz suggested in #12396 . If we want to compare the vector and scalar versions, which script should we run in lucene-util?

Hi,
The problem is you want three variants:
- old index format (scalar)
- new index format scalar
- new index format vectorized 

For the old vs new scalar the indexes are different this needs special handling. Luceneutil shares by default the same index. Maybe you can also compare different indexes, but I don't know how. @mikemccand knows more.

To compare vectorized and scalar new variant use same checkout directory in config and only pass the vector incubator command line in one config and default command line in the other.

Be sure to download a large index (wikimedium at least) and configure it to run enough queries!

If the performance tests are fine we need to cleanup here: make a copy of old codec on backwards codecs module. Rename current codec with your changes to Lucene98 and als rename the ForUtil classes (90 to 98).

This is a bit work but

gsmiller commented:
[Disclaimer: I'm pretty new to working with the incubator vector code, so please excuse wonky terminology or silly ideas]

@ChrisHegarty have you experimented at all with narrower vector lanes for lower bpv? I poked around your bitpacking repo a bit this morning and experimented a bit with the 2bpv unpacking case. If I'm understanding correctly, you essentially have four 32-bit lanes involved in the shift/mask operations, so you're getting 4x parallelism right? (i.e., you're "generating" four output ints in each vectorized op?). I wonder if there's any way to use 8bit lanes and get 16x parallelism in each shift/mask op? I messed about with this a little bit, and—if I unpack into a byte[] instead of int[]—this produced a pretty nice improvement in the benchmarks[1]. What I'm not sure of is how to efficiently get into an int[] for output. Maybe that's the killer? Anyway, wanted to just float this question out there as a newbie to the space but curious and trying to learn more.

[1] here's what I experimented with:
```
    private static final VectorSpecies<Byte> BYTE_SPECIES_128 = ByteVector.SPECIES_128;
    [...]
    byte[] input = [...] // input is in bytes in this case, not int[]
    byte[] byteOut = new byte[128];
    ByteVector inVec = ByteVector.fromArray(BYTE_SPECIES_128, input, 0);
    ByteVector outVec;
    int outOff = 0;
    final byte mask = (1 << 2) - 1;

    outVec = inVec.and(mask);
    outVec.intoArray(byteOut, outOff);

    outVec = inVec.lanewise(VectorOperators.LSHR, 2).and(mask);
    outVec.intoArray(byteOut, outOff+=16);

    outVec = inVec.lanewise(VectorOperators.LSHR, 4).and(mask);
    outVec.intoArray(byteOut, outOff+=16);

    [... you get the idea ...]
```

```
Benchmark                    Mode  Cnt   Score   Error   Units
Benchmark.decode2SimdPack   thrpt    5  58.689 ± 0.191  ops/us
Benchmark.decode2SimdPack2  thrpt    5  85.847 ± 1.800  ops/us
```

tang-hi committed:
fix decodeTo32's bug
the output longs may be reused, so it will have some rubbish value
should use assign instead of bit operation

tang-hi commented:
> What I'm not sure of is how to efficiently get into an int[] for output

Perhaps you could consider utilizing the [reinterprestAsInts] (https://docs.oracle.com/en/java/javase/17/docs/api/jdk.incubator.vector/jdk/incubator/vector/ByteVector.html#reinterpretAsInts()) method? However, I am uncertain if it would be the most optimal approach for this task.

tang-hi committed:
clean commented code

tang-hi commented:
> > What I'm not sure of is how to efficiently get into an int[] for output
> 
> Perhaps you could consider utilizing the [reinterprestAsInts](https://docs.oracle.com/en/java/javase/17/docs/api/jdk.incubator.vector/jdk/incubator/vector/ByteVector.html#reinterpretAsInts()) method? However, I am uncertain if it would be the most optimal approach for this task.

Sorry, perhaps I misunderstood your question. Are you asking how to  convert a byte vector with 16 elements into an array of integers with 16 elements?

tang-hi commented:
I tested the code for vectors and scalars on wikimediumall. Comparing them to the baseline, it shows that neither of them performs better than the baseline. I found that the bottleneck in performance is not in the encoding and decoding, but rather in the prefix sum calculation. The baseline code uses some tricks, such as using a long to store two ints, which allows it to calculate the prefix sum faster. I will post the performance test results below. I have to say that the implementation of PforUtil in the baseline is indeed very clever. My question now is whether we should keep the original compression format or continue seeking a faster prefix sum. Because currently, I don't have any optimization ideas for the current PR. If you have any good ideas, please feel free to suggest.

tang-hi commented:
scalar
  
                          TaskQPS baseline      StdDevQPS my_modified_version      StdDev                Pct diff p-value
                      AndHighLow      444.15      (9.7%)      252.28      (4.2%)  -43.2% ( -52% -  -32%) 0.000
                 LowSloppyPhrase       47.43      (8.0%)       27.70      (4.6%)  -41.6% ( -50% -  -31%) 0.000
                 MedSloppyPhrase       29.58      (4.9%)       17.36      (5.4%)  -41.3% ( -49% -  -32%) 0.000
         AndHighMedDayTaxoFacets       19.57      (6.6%)       11.85      (1.4%)  -39.4% ( -44% -  -33%) 0.000
             MedIntervalsOrdered       18.07      (9.8%)       11.56      (3.5%)  -36.0% ( -44% -  -25%) 0.000
             LowIntervalsOrdered       13.32      (5.5%)        8.59      (4.4%)  -35.5% ( -43% -  -27%) 0.000
                     LowSpanNear        5.46     (10.8%)        3.71      (4.6%)  -32.1% ( -42% -  -18%) 0.000
                       LowPhrase       73.28      (8.6%)       50.64      (4.5%)  -30.9% ( -40% -  -19%) 0.000
                    HighSpanNear       11.49      (8.2%)        8.00      (3.3%)  -30.4% ( -38% -  -20%) 0.000
                      AndHighMed       59.41      (7.0%)       41.40      (4.9%)  -30.3% ( -39% -  -19%) 0.000
                       OrHighMed       99.06      (7.5%)       70.25      (2.8%)  -29.1% ( -36% -  -20%) 0.000
        AndHighHighDayTaxoFacets        5.12      (9.3%)        3.64      (3.1%)  -28.9% ( -37% -  -18%) 0.000
                    OrNotHighLow      754.62      (7.4%)      543.79      (4.8%)  -27.9% ( -37% -  -17%) 0.000
            HighIntervalsOrdered       12.32      (9.2%)        9.06      (4.3%)  -26.5% ( -36% -  -14%) 0.000
                      OrHighHigh       21.81      (7.7%)       16.20      (5.6%)  -25.8% ( -36% -  -13%) 0.000
                       MedPhrase       83.87      (2.8%)       62.51      (2.8%)  -25.5% ( -30% -  -20%) 0.000
                     AndHighHigh       31.21      (7.4%)       23.69      (5.4%)  -24.1% ( -34% -  -12%) 0.000
                HighSloppyPhrase       12.34      (9.1%)        9.67      (5.6%)  -21.6% ( -33% -   -7%) 0.000
                       OrHighLow      296.60      (9.6%)      237.08      (6.0%)  -20.1% ( -32% -   -4%) 0.000
                     MedSpanNear        1.82     (10.8%)        1.48      (7.1%)  -18.9% ( -33% -   -1%) 0.000
                    OrNotHighMed      785.56      (6.6%)      647.04      (5.6%)  -17.6% ( -28% -   -5%) 0.000
                      HighPhrase       45.65      (8.8%)       38.19      (5.6%)  -16.3% ( -28% -   -2%) 0.000
           HighTermDayOfYearSort      294.35      (9.4%)      255.97      (7.3%)  -13.0% ( -27% -    4%) 0.000
                   OrNotHighHigh      499.12      (5.9%)      451.00      (7.1%)   -9.6% ( -21% -    3%) 0.000
                      TermDTSort      163.93     (10.6%)      149.58      (7.6%)   -8.8% ( -24% -   10%) 0.003
               HighTermTitleSort      130.69     (11.9%)      119.57      (6.3%)   -8.5% ( -23% -   10%) 0.005
                   OrHighNotHigh      496.97      (9.0%)      455.82      (6.8%)   -8.3% ( -22% -    8%) 0.001
                         MedTerm      737.16      (9.3%)      676.34      (5.8%)   -8.3% ( -21% -    7%) 0.001
                    OrHighNotMed      554.46      (5.1%)      509.67      (3.5%)   -8.1% ( -15% -    0%) 0.000
          OrHighMedDayTaxoFacets        3.11     (10.6%)        2.86      (5.8%)   -7.9% ( -21% -    9%) 0.003
                         LowTerm      695.06      (7.9%)      647.41      (7.5%)   -6.9% ( -20% -    9%) 0.005
                    OrHighNotLow      501.56     (10.3%)      469.90      (8.0%)   -6.3% ( -22% -   13%) 0.030
            MedTermDayTaxoFacets       40.34      (7.8%)       38.20      (6.0%)   -5.3% ( -17% -    9%) 0.015
                        HighTerm      462.02     (10.9%)      438.96      (4.1%)   -5.0% ( -18% -   11%) 0.056
                          Fuzzy1       54.38      (7.9%)       51.72      (7.2%)   -4.9% ( -18% -   11%) 0.040
                         Prefix3       82.95      (6.4%)       80.20      (7.9%)   -3.3% ( -16% -   11%) 0.145
            HighTermTitleBDVSort       16.28      (3.0%)       15.76      (6.2%)   -3.2% ( -11% -    6%) 0.040
       BrowseDayOfYearTaxoFacets        6.04     (15.1%)        5.87     (14.4%)   -2.7% ( -28% -   31%) 0.563
            BrowseDateTaxoFacets        5.95     (16.7%)        5.82     (16.2%)   -2.3% ( -30% -   36%) 0.658
           BrowseMonthSSDVFacets        9.10     (11.7%)        9.07     (12.3%)   -0.4% ( -21% -   26%) 0.924
               HighTermMonthSort     3206.89      (6.9%)     3198.58      (6.4%)   -0.3% ( -12% -   13%) 0.902
                          IntNRQ       34.93      (8.9%)       34.86      (9.4%)   -0.2% ( -17% -   19%) 0.943
     BrowseRandomLabelTaxoFacets        5.57     (15.8%)        5.56      (9.9%)   -0.2% ( -22% -   30%) 0.964
                          Fuzzy2      123.89      (6.9%)      124.11      (7.4%)    0.2% ( -13% -   15%) 0.935
                         Respell       69.43      (7.9%)       69.81      (7.5%)    0.6% ( -13% -   17%) 0.820
                        Wildcard       58.65      (9.7%)       59.69      (8.9%)    1.8% ( -15% -   22%) 0.548
     BrowseRandomLabelSSDVFacets        5.85     (12.8%)        5.96     (13.5%)    1.9% ( -21% -   32%) 0.649
                        PKLookup      253.11      (9.7%)      258.33      (9.1%)    2.1% ( -15% -   23%) 0.488
       BrowseDayOfYearSSDVFacets        7.99     (17.0%)        8.16     (12.3%)    2.1% ( -23% -   37%) 0.647
            BrowseDateSSDVFacets        1.62     (18.9%)        1.66     (14.7%)    2.7% ( -26% -   44%) 0.619
           BrowseMonthTaxoFacets       12.29     (46.5%)       12.68     (47.1%)    3.2% ( -61% -  180%) 0.830

tang-hi commented:
vectorized

                           TaskQPS baseline      StdDevQPS my_modified_version      StdDev                Pct diff p-value
                         Prefix3      195.72      (3.6%)      152.81      (2.6%)  -21.9% ( -27% -  -16%) 0.000
                        Wildcard      140.47      (4.1%)      124.83      (3.4%)  -11.1% ( -17% -   -3%) 0.000
                    OrNotHighLow      530.23      (3.4%)      495.25      (5.5%)   -6.6% ( -14% -    2%) 0.000
            BrowseDateTaxoFacets        5.84     (18.5%)        5.50      (8.8%)   -5.8% ( -27% -   26%) 0.208
                      AndHighLow      925.55      (4.0%)      872.82      (3.6%)   -5.7% ( -12% -    1%) 0.000
       BrowseDayOfYearTaxoFacets        5.91     (18.6%)        5.58      (8.7%)   -5.5% ( -27% -   26%) 0.229
                          Fuzzy2       88.72      (3.6%)       84.66      (4.8%)   -4.6% ( -12% -    3%) 0.001
                    OrHighNotLow      499.42      (5.0%)      478.58      (5.3%)   -4.2% ( -13% -    6%) 0.010
            BrowseDateSSDVFacets        1.85      (9.0%)        1.80      (8.6%)   -2.8% ( -18% -   16%) 0.311
               HighTermTitleSort       51.91      (5.7%)       50.58      (4.5%)   -2.6% ( -12% -    8%) 0.112
                       OrHighLow      424.81      (3.5%)      413.94      (3.3%)   -2.6% (  -9% -    4%) 0.017
           HighTermDayOfYearSort      250.13      (2.7%)      244.81      (3.7%)   -2.1% (  -8% -    4%) 0.039
     BrowseRandomLabelTaxoFacets        5.15      (9.4%)        5.06      (8.6%)   -1.8% ( -18% -   17%) 0.522
           BrowseMonthTaxoFacets       18.04      (3.8%)       17.74     (12.3%)   -1.7% ( -17% -   14%) 0.560
            MedTermDayTaxoFacets       13.70      (3.8%)       13.52      (4.3%)   -1.3% (  -9% -    7%) 0.300
          OrHighMedDayTaxoFacets        6.69      (5.0%)        6.64      (4.8%)   -0.8% ( -10% -    9%) 0.614
                     LowSpanNear       41.13      (3.7%)       40.85      (3.5%)   -0.7% (  -7% -    6%) 0.551
                      TermDTSort      151.24      (4.0%)      150.29      (4.8%)   -0.6% (  -9% -    8%) 0.655
                          Fuzzy1       78.36      (3.8%)       78.19      (2.9%)   -0.2% (  -6% -    6%) 0.840
        AndHighHighDayTaxoFacets       13.65      (2.6%)       13.64      (3.5%)   -0.1% (  -6% -    6%) 0.936
                     MedSpanNear        5.89      (4.9%)        5.90      (3.0%)    0.1% (  -7% -    8%) 0.932
       BrowseDayOfYearSSDVFacets        6.24      (5.8%)        6.25      (5.2%)    0.2% ( -10% -   11%) 0.918
                   OrHighNotHigh      388.28      (4.9%)      389.16      (5.5%)    0.2% (  -9% -   11%) 0.890
                    OrNotHighMed      445.39      (3.7%)      446.57      (3.5%)    0.3% (  -6% -    7%) 0.817
                         MedTerm      658.90      (4.0%)      660.78      (4.7%)    0.3% (  -8% -    9%) 0.837
                    OrHighNotMed      538.54      (4.3%)      541.11      (5.6%)    0.5% (  -9% -   10%) 0.763
           BrowseMonthSSDVFacets        6.34      (3.4%)        6.38      (6.5%)    0.6% (  -8% -   10%) 0.715
                         Respell       71.93      (3.8%)       72.42      (3.1%)    0.7% (  -5% -    7%) 0.530
                    HighSpanNear       15.68      (4.3%)       15.81      (3.7%)    0.8% (  -6% -    9%) 0.522
         AndHighMedDayTaxoFacets       10.22      (3.6%)       10.34      (4.2%)    1.2% (  -6% -    9%) 0.327
                         LowTerm      732.95      (3.7%)      742.78      (3.7%)    1.3% (  -5% -    9%) 0.250
                      HighPhrase       42.41      (3.3%)       43.03      (3.3%)    1.4% (  -4% -    8%) 0.161
            HighTermTitleBDVSort       17.18      (4.0%)       17.44      (3.4%)    1.5% (  -5% -    9%) 0.203
                   OrNotHighHigh      361.98      (4.1%)      367.56      (4.5%)    1.5% (  -6% -   10%) 0.255
                      OrHighHigh       20.00      (5.3%)       20.31      (7.4%)    1.6% ( -10% -   15%) 0.435
                        HighTerm      549.87      (4.4%)      558.76      (4.4%)    1.6% (  -6% -   10%) 0.243
                          IntNRQ       41.31      (8.4%)       42.09      (6.8%)    1.9% ( -12% -   18%) 0.440
     BrowseRandomLabelSSDVFacets        4.96      (6.2%)        5.06      (8.0%)    1.9% ( -11% -   17%) 0.398
                        PKLookup      260.58      (5.6%)      265.62      (3.9%)    1.9% (  -7% -   12%) 0.204
                     AndHighHigh       20.02      (4.0%)       20.44      (4.2%)    2.1% (  -5% -   10%) 0.110
                 LowSloppyPhrase       52.29      (4.4%)       53.44      (4.1%)    2.2% (  -6% -   11%) 0.104
               HighTermMonthSort     2970.91      (6.7%)     3036.52      (6.5%)    2.2% ( -10% -   16%) 0.291
             LowIntervalsOrdered        1.95      (5.2%)        2.00      (4.0%)    2.2% (  -6% -   12%) 0.130
                       MedPhrase      185.75      (3.8%)      189.93      (2.4%)    2.2% (  -3% -    8%) 0.026
                 MedSloppyPhrase       48.01      (4.6%)       49.13      (5.1%)    2.3% (  -7% -   12%) 0.129
                       OrHighMed       81.95      (4.7%)       84.09      (4.7%)    2.6% (  -6% -   12%) 0.080
            HighIntervalsOrdered       14.29      (5.2%)       14.71      (3.6%)    2.9% (  -5% -   12%) 0.040
                      AndHighMed      104.16      (5.0%)      107.62      (4.9%)    3.3% (  -6% -   13%) 0.033
                       LowPhrase      232.12      (4.2%)      239.90      (3.0%)    3.4% (  -3% -   11%) 0.004
             MedIntervalsOrdered        6.13      (5.3%)        6.38      (4.0%)    4.1% (  -4% -   14%) 0.006
                HighSloppyPhrase        2.43      (5.5%)        2.53      (4.6%)    4.1% (  -5% -   15%) 0.011

ChrisHegarty commented:
> Comparing them to the baseline, it shows that neither of them performs better than the baseline.

This is a little surprising, and also disappointing. I had assumed, without verifying, that there was significant overhead in the bit packing. I'll have to spend some more time on the benchmark results.

> I found that the bottleneck in performance is not in the encoding and decoding, but rather in the prefix sum calculation. 

Hmm.. @jpountz and I have been looking a little at this, see [vectorized-prefix-sum][1]. But there are issues there, with how the Vector API handles cross-lane moves - they are not optimal. I've yet to get this feedback to OpenJDK.

I think we should step back a little here, and get a better handle on exactly where the most significant bottleneck is. @tang-hi you say it is in prefix sum. How are you determining that?

Some background on the current implementation: [LUCENE-9027: Use SIMD instructions to decode postings](https://issues.apache.org/jira/browse/LUCENE-9027)

[1]: https://github.com/jpountz/vectorized-prefix-sum

tang-hi commented:
> you say it is in prefix sum. How are you determining that?

When I noticed a significant decrease in performance even with vectorized code, I reexamined the benchmark. Surprisingly, despite the significant performance advantages of packing and unpacking, the baseline performance was better for `decodeTo32`. Upon investigation, I found that its compression format allows it to conveniently represent two ints using a single long, resulting in only 64 iterations needed for prefix sum calculation. This convenience does not apply to our modified compression format, so I attempted to change it to directly calculate the prefix sum after decoding (and fully unrolling it). After running the benchmark, I found that the performance remained relatively the same.

tang-hi commented:
If there is a need for me to provide assistance in any subsequent performance testing, please feel free to let me know, such as the built index (building an index can be quite time-consuming).

uschindler commented:
Hi,
what I figured out with performance testing is the following: When you use default settings it doe snot run enough queries by default and instead repeats the whole JVM startup several times (i think 20 times). Because the optimizations of hotspot are quite slow and at beginning the code creates many object instances on heap so GC runs extensively, so you need to tune the settings a bit:
- use huge dataset (wikimedium is ok, but needs more queries)
- use many queries per run
- reduce number of JVM startups (5 should be enough)

Uwe

ChrisHegarty commented:
Part of the issue here is that we are (well, at least I have been), looking at bit packing and unpacking in isolation. We really need to consider the impact on the codec more holistically - including prefix sum (which we've not done so far). It's silly to represent the decoded values as int[], and then coerce then into long[] for prefix sum - we should just vectorise prefix sum to work on int[] (in a reasonable way).  I don't even see how this is supposed to work in this PR - have I missed something ?

tang-hi commented:
> It's silly to represent the decoded values as int[], and then coerce then into long[] for prefix sum - we should just vectorise prefix sum to work on int[] (in a reasonable way)

Yes, I realized this issue and then tried to directly sum the int[] using a simple for loop. Comparing it with the baseline using jmh, the effect doesn't seem to be much better. Of course, I haven't tested how much improvement there will be in luceneutil's bench after making this change. Currently, I have only found two places that need to be modified, namely the two occurrences of decodeTo32 in PForUtil. Maybe later we can directly sum and then see what the benchmark looks like.

gsmiller commented:
> Hi! I have an idea. Since our encode and decode functions are fast, can we lazy compute prefix sums? We can first decompress them and only calculate the actual values when the user needs them. This might save some time. It's just a preliminary idea, and I would like to hear everyone's opinions. If everyone thinks it's a good idea, I might try implementing it and see how it works.

I like the idea of trying this out, but maybe in a spin-off issue? @rmuir raised it a little while back as well (need to dig up a link to his comment). I think it may add additional complication to getting this right since the performance will start to depend a lot on how "sparse" a query is and where the matching docs fall in a block?

uschindler commented:
> @uschindler Understood, I will adjust the settings. But first, I need to go and find where the settings are located.😀

From my experience with MMapDircetory, you should be fine if the benchmark takes about a minimum of 2 minutes per round, so 40 minutes in total (unless you reduce number of iterations). So it is better to have less iterations and longer runtime.

gsmiller commented:
> @gsmiller It is not surprising that ByteVector is faster for narrow bit width scenarios. The issue is how to represent these different bit widths in a cohesive way, without coercing the bits into a narrower format - since that defeats the point.

@ChrisHegarty I'm not entirely sure what you mean here, but I'm certain that's my lack of understanding in this space surfacing.

Just taking a bit width of 2 as an illustrative example, my super-naive thought is that we'd be encoding the values to disk using essentially 8 total lanes right? So we'd store 8 ints to represent these 8 lanes (each int packing 16 2-bit values)? The first int would store docs 0, 8, 16, 32, [...], 120 (in the block of 128 docs). The second int would store 1, 9, 17, 33, [...], 121. And so on. For each of these ints, we need to perform 16 "operations" to decode the 16 values (shift + mask). If we're using a 128 bit vector register, we can operate on 4 lanes at a time, so we have to load up 4 ints, do these 16 operations, then do the next 4, etc. But then each lane directly contains the int we want to "copy out" into the final int[] we're populating, so that part is easy.

Alternatively, what if we used 32 lanes that were only 8 bits wide instead of 8 lanes that are 32 bits wide? If we stored 32 bytes to represent these lanes, each packing 4 values, we'd have something like lane 1 containing 0, 32, 64, 96; lane 2 containing 1, 33, 65, 97; and so on up to our last lane of 31, 63, 95, 127. We could read these bytes from disk and directly load 16 of them into a 128 bit register, performing 16 concurrent shift + mask operations then copying out 16 values at a time instead of only 4 as before. What I'm not sure about though is whether-or-not it's possible to copy out those 16 values at a time into ints. We really just care about copying out the 8 bit lane values and left-padding the remaining 24 bits with 0s, but maybe that can't easily be done? I messed about with `reinterpretAsInts`, but that doesn't really seem well suited? Maybe there's a better way to approach this? Or maybe it just fundamentally can't be done?

I clearly have a large knowledge gap in this space, so again—I'm sure this reasoning is fundamentally flawed in some way—but just approaching it with a bit of an outsider's perspective and wondering why we can't increase our concurrency here in cases of small bit widths. Just seems a bit wasteful to shift + mask 128 bits at a time and only care about 8 of them when we could shift + mask 128 bits at a time and care about 32 of them.

Also, happy to take this conversation elsewhere if it makes sense. If it goes beyond this comment, I'll move it somewhere else to make sure we don't clutter up this PR. Thanks for your help understanding!

tang-hi commented:
Hi, everyone. I tried the lazy compute idea that I mentioned before. First, I attempted to change the code in the main branch to lazy compute,  the benchmark results didn't show much difference. Then I applied the lazy compute algorithm to the vectorized code, and the benchmark results showed improved performance. However, I was surprised to see that the benchmark results for Prefix3 were not good. After that, I tested the scalar code, and the benchmark results showed a decrease in performance. I have the results of these two benchemark below.

## vectorized

                            TaskQPS baseline      StdDevQPS my_modified_version      StdDev                Pct diff p-value
                         Prefix3       90.71      (2.4%)       80.20      (4.7%)  -11.6% ( -18% -   -4%) 0.000
       BrowseDayOfYearSSDVFacets        6.89     (13.1%)        6.65     (10.2%)   -3.5% ( -23% -   22%) 0.348
                        Wildcard       56.40      (3.0%)       54.70      (2.9%)   -3.0% (  -8% -    3%) 0.001
            BrowseDateTaxoFacets        5.72     (11.1%)        5.58      (7.9%)   -2.5% ( -19% -   18%) 0.415
     BrowseRandomLabelTaxoFacets        5.54      (8.6%)        5.42      (7.4%)   -2.3% ( -16% -   14%) 0.369
       BrowseDayOfYearTaxoFacets        5.73     (10.3%)        5.61      (7.5%)   -2.1% ( -18% -   17%) 0.454
           BrowseMonthSSDVFacets        7.01     (11.3%)        6.86     (11.9%)   -2.1% ( -22% -   23%) 0.567
                          IntNRQ       42.16      (4.9%)       41.72      (2.7%)   -1.1% (  -8% -    6%) 0.397
                          Fuzzy2       81.88      (1.8%)       82.32      (2.3%)    0.5% (  -3% -    4%) 0.411
             MedIntervalsOrdered       10.00      (5.2%)       10.05      (4.0%)    0.5% (  -8% -   10%) 0.711
                          Fuzzy1      114.12      (1.8%)      114.78      (2.4%)    0.6% (  -3% -    4%) 0.385
     BrowseRandomLabelSSDVFacets        5.31      (5.5%)        5.35      (8.0%)    0.7% ( -12% -   14%) 0.746
                      TermDTSort      159.43      (5.3%)      160.63      (4.0%)    0.8% (  -8% -   10%) 0.612
            MedTermDayTaxoFacets       11.07      (3.0%)       11.16      (2.8%)    0.8% (  -4% -    6%) 0.381
               HighTermTitleSort      100.01      (3.8%)      101.02      (4.6%)    1.0% (  -7% -    9%) 0.447
                     MedSpanNear       45.12      (2.5%)       45.61      (2.3%)    1.1% (  -3% -    6%) 0.157
            HighIntervalsOrdered       14.06      (4.6%)       14.22      (4.3%)    1.2% (  -7% -   10%) 0.410
            HighTermTitleBDVSort       12.39      (1.7%)       12.54      (2.9%)    1.2% (  -3% -    5%) 0.104
                    HighSpanNear       17.67      (2.2%)       17.90      (2.0%)    1.3% (  -2% -    5%) 0.047
                         Respell       68.69      (1.5%)       69.62      (1.5%)    1.4% (  -1% -    4%) 0.005
            BrowseDateSSDVFacets        1.78     (13.0%)        1.81     (11.8%)    1.6% ( -20% -   30%) 0.676
        AndHighHighDayTaxoFacets        3.79      (3.4%)        3.86      (3.4%)    2.0% (  -4% -    9%) 0.067
         AndHighMedDayTaxoFacets       21.71      (2.5%)       22.16      (2.4%)    2.1% (  -2% -    7%) 0.007
          OrHighMedDayTaxoFacets        8.33      (5.1%)        8.52      (3.5%)    2.3% (  -6% -   11%) 0.103
                        PKLookup      266.99      (2.7%)      273.32      (2.4%)    2.4% (  -2% -    7%) 0.003
                    OrHighNotLow      536.68      (3.5%)      551.31      (4.7%)    2.7% (  -5% -   11%) 0.037
                        HighTerm      772.12      (2.2%)      795.02      (3.7%)    3.0% (  -2% -    9%) 0.002
                         LowTerm      777.24      (2.1%)      803.48      (4.0%)    3.4% (  -2% -    9%) 0.001
                         MedTerm      625.99      (3.1%)      647.55      (5.4%)    3.4% (  -4% -   12%) 0.013
                 MedSloppyPhrase       18.33      (1.8%)       18.98      (2.8%)    3.5% (   0% -    8%) 0.000
                    OrHighNotMed      461.93      (3.6%)      478.49      (4.2%)    3.6% (  -4% -   11%) 0.004
                   OrHighNotHigh      526.32      (2.9%)      546.51      (3.7%)    3.8% (  -2% -   10%) 0.000
                    OrNotHighMed      417.97      (2.6%)      434.74      (3.1%)    4.0% (  -1% -    9%) 0.000
                   OrNotHighHigh      514.95      (2.9%)      535.68      (2.7%)    4.0% (  -1% -    9%) 0.000
                      OrHighHigh       37.60      (3.5%)       39.18      (4.5%)    4.2% (  -3% -   12%) 0.001
                HighSloppyPhrase        2.28      (2.3%)        2.38      (2.4%)    4.3% (   0% -    9%) 0.000
                     AndHighHigh       38.76      (2.1%)       40.63      (2.8%)    4.8% (   0% -    9%) 0.000
           HighTermDayOfYearSort      343.83      (2.4%)      360.72      (4.7%)    4.9% (  -2% -   12%) 0.000
                 LowSloppyPhrase       60.08      (1.6%)       63.04      (2.2%)    4.9% (   1% -    8%) 0.000
                    OrNotHighLow      647.85      (1.7%)      680.90      (2.5%)    5.1% (   0% -    9%) 0.000
             LowIntervalsOrdered        5.71      (3.6%)        6.00      (2.8%)    5.1% (  -1% -   11%) 0.000
                      HighPhrase       28.70      (1.7%)       30.18      (1.7%)    5.2% (   1% -    8%) 0.000
                     LowSpanNear        7.41      (2.1%)        7.80      (1.9%)    5.3% (   1% -    9%) 0.000
                       MedPhrase       11.08      (1.5%)       11.71      (2.1%)    5.8% (   2% -    9%) 0.000
                       OrHighLow      520.52      (1.6%)      552.94      (3.5%)    6.2% (   1% -   11%) 0.000
                      AndHighLow     1225.95      (2.7%)     1308.18      (5.8%)    6.7% (  -1% -   15%) 0.000
               HighTermMonthSort     3123.59      (2.9%)     3381.95      (5.3%)    8.3% (   0% -   16%) 0.000
                       LowPhrase       27.73      (1.3%)       30.05      (2.2%)    8.4% (   4% -   12%) 0.000
                      AndHighMed      109.92      (1.8%)      119.78      (2.7%)    9.0% (   4% -   13%) 0.000
                       OrHighMed      111.14      (2.2%)      121.78      (3.3%)    9.6% (   3% -   15%) 0.000
           BrowseMonthTaxoFacets       16.77     (28.5%)       18.90      (1.4%)   12.7% ( -13% -   59%) 0.046

## scalar

                            TaskQPS baseline      StdDevQPS my_modified_version      StdDev                Pct diff p-value
                 LowSloppyPhrase        9.77      (4.4%)        5.73      (2.1%)  -41.4% ( -45% -  -36%) 0.000
                     MedSpanNear       11.23      (4.6%)        6.95      (2.3%)  -38.2% ( -43% -  -32%) 0.000
                      AndHighMed      123.37      (8.0%)       76.65      (2.0%)  -37.9% ( -44% -  -30%) 0.000
         AndHighMedDayTaxoFacets       26.15      (2.0%)       16.34      (1.6%)  -37.5% ( -40% -  -34%) 0.000
                    HighSpanNear       13.97      (3.2%)        9.09      (1.8%)  -34.9% ( -38% -  -30%) 0.000
             MedIntervalsOrdered       21.95      (3.5%)       14.81      (2.7%)  -32.5% ( -37% -  -27%) 0.000
                      AndHighLow     1070.91      (3.9%)      732.57      (2.7%)  -31.6% ( -36% -  -25%) 0.000
                       LowPhrase       20.41      (2.9%)       14.15      (2.3%)  -30.7% ( -34% -  -26%) 0.000
                 MedSloppyPhrase       28.12      (2.8%)       19.93      (1.2%)  -29.1% ( -32% -  -25%) 0.000
             LowIntervalsOrdered        3.36      (4.6%)        2.39      (3.3%)  -28.8% ( -35% -  -21%) 0.000
                    OrNotHighLow      640.26      (2.8%)      488.06      (1.8%)  -23.8% ( -27% -  -19%) 0.000
        AndHighHighDayTaxoFacets       10.74      (2.1%)        8.19      (2.1%)  -23.7% ( -27% -  -19%) 0.000
                     LowSpanNear      277.31      (1.8%)      212.48      (1.9%)  -23.4% ( -26% -  -20%) 0.000
                     AndHighHigh       25.31      (5.6%)       19.62      (2.7%)  -22.5% ( -29% -  -15%) 0.000
                    OrNotHighMed      528.99      (2.2%)      414.93      (2.7%)  -21.6% ( -25% -  -17%) 0.000
                       OrHighLow      408.04      (3.1%)      322.73      (2.7%)  -20.9% ( -25% -  -15%) 0.000
                HighSloppyPhrase       13.08      (3.5%)       10.65      (2.0%)  -18.6% ( -23% -  -13%) 0.000
                      OrHighHigh       23.91      (5.3%)       19.49      (2.5%)  -18.5% ( -24% -  -11%) 0.000
                       MedPhrase      187.62      (2.3%)      153.52      (1.7%)  -18.2% ( -21% -  -14%) 0.000
                       OrHighMed       31.30      (4.8%)       26.68      (3.4%)  -14.8% ( -21% -   -6%) 0.000
            HighIntervalsOrdered        0.74      (5.2%)        0.64      (4.7%)  -14.1% ( -22% -   -4%) 0.000
           HighTermDayOfYearSort      327.16      (2.4%)      288.14      (3.1%)  -11.9% ( -16% -   -6%) 0.000
               HighTermTitleSort      105.28      (6.8%)       93.57      (5.1%)  -11.1% ( -21% -    0%) 0.000
                      HighPhrase      173.78      (2.5%)      156.10      (1.4%)  -10.2% ( -13% -   -6%) 0.000
                      TermDTSort      166.57      (5.5%)      151.19      (4.4%)   -9.2% ( -18% -    0%) 0.000
                         MedTerm      641.21      (3.4%)      583.88      (3.2%)   -8.9% ( -15% -   -2%) 0.000
                   OrNotHighHigh      533.02      (2.8%)      486.18      (2.0%)   -8.8% ( -13% -   -4%) 0.000
                    OrHighNotLow      516.66      (2.9%)      472.41      (4.7%)   -8.6% ( -15% -    0%) 0.000
          OrHighMedDayTaxoFacets        8.58      (3.7%)        7.85      (3.8%)   -8.4% ( -15% -    0%) 0.000
                    OrHighNotMed      496.57      (3.1%)      457.70      (2.7%)   -7.8% ( -13% -   -2%) 0.000
                        HighTerm      587.86      (4.2%)      542.76      (3.0%)   -7.7% ( -14% -    0%) 0.000
                   OrHighNotHigh      770.22      (2.5%)      723.29      (2.9%)   -6.1% ( -11% -    0%) 0.000
                          Fuzzy2       31.75      (2.2%)       29.98      (5.6%)   -5.6% ( -13% -    2%) 0.000
            MedTermDayTaxoFacets       38.41      (1.8%)       36.45      (1.4%)   -5.1% (  -8% -   -1%) 0.000
                         LowTerm      900.74      (2.3%)      866.32      (3.2%)   -3.8% (  -9% -    1%) 0.000
            HighTermTitleBDVSort        7.78      (2.9%)        7.61      (3.6%)   -2.2% (  -8% -    4%) 0.034
           BrowseMonthSSDVFacets        9.22      (6.5%)        9.08      (7.8%)   -1.4% ( -14% -   13%) 0.532
                        PKLookup      260.37      (3.8%)      258.96      (3.6%)   -0.5% (  -7% -    7%) 0.645
                         Respell       54.02      (2.2%)       53.76      (2.2%)   -0.5% (  -4% -    4%) 0.492
                          Fuzzy1       81.60      (2.4%)       81.29      (1.7%)   -0.4% (  -4% -    3%) 0.557
     BrowseRandomLabelSSDVFacets        6.07      (8.1%)        6.05      (8.3%)   -0.3% ( -15% -   17%) 0.894
       BrowseDayOfYearSSDVFacets        8.14      (6.9%)        8.12      (7.3%)   -0.3% ( -13% -   14%) 0.885
               HighTermMonthSort     2956.82      (4.0%)     2967.31      (4.3%)    0.4% (  -7% -    9%) 0.787
                        Wildcard       73.82      (3.0%)       75.25      (3.5%)    1.9% (  -4% -    8%) 0.061
                         Prefix3      190.39      (4.0%)      195.04      (3.9%)    2.4% (  -5% -   10%) 0.050
     BrowseRandomLabelTaxoFacets        5.77     (11.3%)        5.92     (10.1%)    2.6% ( -16% -   27%) 0.441
            BrowseDateTaxoFacets        6.26     (15.9%)        6.43     (15.7%)    2.7% ( -24% -   40%) 0.588
       BrowseDayOfYearTaxoFacets        6.28     (15.9%)        6.47     (16.0%)    2.9% ( -25% -   41%) 0.565
           BrowseMonthTaxoFacets       10.24     (52.4%)       10.54     (54.2%)    2.9% ( -68% -  230%) 0.862
                          IntNRQ       43.93     (15.3%)       45.33      (8.8%)    3.2% ( -18% -   32%) 0.419
            BrowseDateSSDVFacets        1.57     (11.2%)        1.64     (13.6%)    4.6% ( -18% -   33%) 0.244


My current conclusions are:
1. Lazy compute works well when combined with vectors.
2. Perhaps we can improve the performance of the scalar code. The current version may still have room for improvement, but I currently don't have any ideas. If you have any suggestions, please feel free to propose them or directly commit to this branch.
3. The benchmark results for Prefix3 seem odd because it only shows a performance decrease in the vectorized version, even in the scalar version, the performance of Prefix3 doesn't decrease. Any clues?

I would love to hear your opinions and welcome any ideas.

gsmiller commented:
Those are interesting results @tang-hi. Thanks for sharing! I also notice a regression in `wildcard` in the vectorized approach as well. The prefix3 and wildcard queries both build an up-front bitset representing all the matching hits, so they're probably getting more benefit from bulk decoding of the docIds vs. a lazy solution. It's curious that there isn't a corresponding regression in your scalar solution as well.

gsmiller commented:
OK, one last follow-up on my idea of using tighter lanes to get more concurrency for shift+mask while decoding: I educated myself a little bit more and explored one last idea of using a scatter instruction to copy the output vector to a byte[] that is padded with 0s such that each decoded value is 4 bytes instead of 1. My thought was that we could potentially throw a ByteBuffer on top of this and interpret the data as integers. From some micro-benchmarks I ran, it looks like the "scatter" is very costly and tanks performance, so I'm convinced at this point it's not worth pursuing the "tighter lane" idea further. Here's essentially what I was doing just to close the loop (I was experimenting with 64 values at a time instead of 128, and a bit width of 2 just to make things a bit simpler and only load data to a vector register once):

```
    private static final int[] scatterMap = new int[16];
    static {
        int upto = 3;
        for (int i = 0; i < 16; i++) {
            scatterMap[i] = upto;
            upto += 4;
        }
    }

    public static void unpackSimd2(byte[] in, byte[] out) {
        ByteVector inVec = ByteVector.fromArray(BYTE_SPECIES_128, in, 0);
        ByteVector outVec;
        int upto = 0;
        for (int shift = 0; shift < 8; shift += 2) {
            outVec = inVec.lanewise(VectorOperators.LSHR, shift).and(BYTE_MASK);
            outVec.intoArray(out, upto, scatterMap, 0);
            upto += 64;
        }
    }
```

gsmiller commented:
Now for something possibly more useful. I brought the current vectorized version of ForUtil from this PR over to the local Lucene fork we use for Amazon product search and ran some internal benchmarks (replaying sampled production traffic against a single-box version of our service). I did this on two different AWS ec2 instances, where the baseline uses the ForUtil implementation as it exists on 9.4, and the candidate drops the vectorized implementation in place (along with the slight tweaks to PForUtil, Lucene90PostingsReader and Lucene90PostingsWriter). Note that these benchmarks were running with jdk19 (not 20), so it's possible we'd see something different with 20?

First, on an x86-based m4.10xlarge instance [1], the vectorized version appears to provide a ~3% red-line QPS improvement along with a 3.7% avg latency reduction.

Next, on an ARM-based (AWS Graviton2) m6g.4xlarge instance[2], I only saw a ~2% red-line QPS improvement and no real avg latency reduction.

I'm going to be mostly away from my computer next week, but I may dive into these benchmarks a bit more after that. Thought it could be useful to put these results out there in the meantime though.

[1] lscpu info for the m4.10xlarge box:
```
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              40
On-line CPU(s) list: 0-39
Thread(s) per core:  2
Core(s) per socket:  10
Socket(s):           2
NUMA node(s):        2
Vendor ID:           GenuineIntel
CPU family:          6
Model:               63
Model name:          Intel(R) Xeon(R) CPU E5-2676 v3 @ 2.40GHz
Stepping:            2
CPU MHz:             2700.156
CPU max MHz:         3000.0000
CPU min MHz:         1200.0000
BogoMIPS:            4800.04
Hypervisor vendor:   Xen
Virtualization type: full
L1d cache:           32K
L1i cache:           32K
L2 cache:            256K
L3 cache:            30720K
NUMA node0 CPU(s):   0-9,20-29
NUMA node1 CPU(s):   10-19,30-39
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq monitor est ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm cpuid_fault invpcid_single pti fsgsbase bmi1 avx2 smep bmi2 erms invpcid xsaveopt ida
```

[2] lscpu info for the m6g.4xlarge box:
```
Architecture:        aarch64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              16
On-line CPU(s) list: 0-15
Thread(s) per core:  1
Core(s) per socket:  16
Socket(s):           1
NUMA node(s):        1
Vendor ID:           ARM
Model:               1
Model name:          Neoverse-N1
Stepping:            r3p1
BogoMIPS:            243.75
L1d cache:           64K
L1i cache:           64K
L2 cache:            1024K
L3 cache:            32768K
NUMA node0 CPU(s):   0-15
Flags:               fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp ssbs
```

tang-hi commented:
It's great to see the performance results in a production environment. May I inquire if you employ the concept of lazy computation? If not, I believe it would be a big  enhancement to incorporate it in conjunction with vectorized code.

ChrisHegarty commented:
Apologies for my tardy and terse interaction here. I've been otherwise preoccupied. I hope to spend time on this soon.

uschindler commented:
> Note that these benchmarks were running with jdk19 (not 20), so it's possible we'd see something different with 20?

Lucene enables and compiles the vectorized code only for jdk 20 and 21. In 19 it won't be enabled.

Be sure to also show the .message logged on startup by the `VectorizationProvider`!

rmuir commented:
please, lets not use this integer vectorization when `hasFastIntegerVectors` is false. Otherwise we can see 30x or so slowdown on virtualmachines without properly plumbed AVX.

tang-hi committed:
compute lazily for post doc

tang-hi commented:
@gsmiller I have pushed the commit for lazy computation.IMO, it will provide better performance. I would highly recommend considering its implementation in the production environment, utilizing JDK version 20 or 21

uschindler commented on lucene/core/src/java20/org/apache/lucene/internal/vectorization/PanamaVectorizationProvider.java:
```diff
@@ -82,4 +82,9 @@ private static <T> T doPrivileged(PrivilegedAction<T> action) {
+
+  @Override
+  public ForUtil90 newForUtil90() {
+    return new PanamaForUtil90();
```
About @rmuir comment:

Please change this to:
```java
return hasFastIntegerVectors ?  new PanamaForUtil90()

tang-hi replied:
done

tang-hi committed:
return DefaultForUtil90 when hasFastIntegerVectors is false

uschindler commented:
> please, lets not use this integer vectorization when `hasFastIntegerVectors` is false. Otherwise we can see 30x or so slowdown on virtualmachines without properly plumbed AVX.

Was a simple change, thanks for reminder! 🖖

rmuir commented:
> Was a simple change, thanks for reminder!

thanks! We can always test performance with `-XX:UseAVX=0`, maybe the check is unneeded based on whatever hotspot can do. Only shifts/and/or seem to be used and maybe these work fine with hotspots SSE support? But this is really no different from the horrible performance we saw with the integer vector math functions so I want us to be careful...

ChrisHegarty commented:
Here's where I'm at, after spending the best part of the last three days hacking in this area - I'm on the fence about whether or not this is worth it.

The current code and format are heavily optimised. The pattern plays to hotspot's auto-vectorization very well, generating AVX 2 instructions for quite a bit of the decoding. Along with the embedding of two integers into a single long, the prefix sum is quite efficient too. This is all to say that the bar is quite high to be "better".

Since Panama is still incubating then we need a scalar fallback. This doesn't necessarily need to be as efficient as the current code, but it should be close (for some definition of close!). I personally find the Panama-ized code more straightforward and easier to reason about, but it requires a new format to work well. This format is superior for the Panama code, but less so for the scalar code.  What bothers me even more is that we cannot easily integrate the prefix sum calculation into the unpack - as we run into Panama bounds check issues that make the performance very poor.

The format proposed here works well for 128 bit widths, but won't for larger. So while I did proposed that we do favour 128, I'm less sure about committing to a new format that we know cannot be optimised further, without of course changing it again.

I tried rewriting parts of the current code, retaining the format, and using Panama. It's possible, but buys us very little (the current code already largely compiles to AVX 2). In fact, heavily mixing AVX 512 for some parts of the decode with scalar for the tail bits results in horrid CPU slow downs. So we cannot even do that to get the "easy" bits of the current code to use AVX 512 rather than AVX2 :-(

tang-hi commented:
I also currently believe that it may not be a good time to vectorize it. Although vectorized code combined with lazy compute does improve performance, we currently cannot achieve scalar code that can match the performance of the current code. Therefore, I think maybe we should wait until Panama is officially released before vectorizing this part of the code.

uschindler commented:
Hi, if Panama is officially released and lucene is using that version as minimum requirement, there is still the problem that you need scalar code for the case where the CPU does not offer all features needed.

We still need to open issues at OpenJDK to fix the problem of slowness when you use vectorized code and the hardware does not support it. Currently we have some hacks to detect at least some features, but there should be ways to detect "hardware supports" for all operators and if C2 is actually enabled.

gsmiller commented:
Was away from my computer for a few days, so just catching up. Will try to respond to a few things here:

gsmiller commented:
@ChrisHegarty I was considering some experimentation with [vectorized prefix sum implementations](https://en.algorithmica.org/hpc/algorithms/prefix/), but saw your comment above stating:

> What bothers me even more is that we cannot easily integrate the prefix sum calculation into the unpack - as we run into Panama bounds check issues that make the performance very poor.

I also came across some [benchmarks](https://github.com/jpountz/vectorized-prefix-sum) it looks like you may have collaborated on with @jpountz related to some different prefix sum SIMD approaches.

Can you elaborate any more on the performance issues related to these vectorized attempts? I assume the benchmark results were poor for you as well (I've tested on a few different machines with pretty horrid results, but I don't really understand why they're so bad).
